### PR TITLE
[STUDIO-4266] Improve mechanism for generate heavy HTML reports

### PR DESCRIPTION
**Pull request by linhtnlK**
[STUDIO-4266](https://katalon.atlassian.net/browse/STUDIO-4266)
More info [comment](https://katalon.atlassian.net/browse/STUDIO-4266?focusedCommentId=132256)

[STUDIO-4266]: https://katalon.atlassian.net/browse/STUDIO-4266?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Include/scripts/groovy/com/kms/katalon/core/reporting/basic/reporting/ReportWriterUtil.java [hunk 1]
```diff
@@ -148,17 +204,19 @@ public static void writeLogRecordToFiles(TestSuiteLogRecord suiteLogEntity, File
 
         writeSimpleHTMLReport(suiteLogEntity, logFolder);
 
-//        writeJsonReport(suiteLogEntity, logFolder);
+        // writeJsonReport(suiteLogEntity, logFolder);
```
Comment (Minh-Chien) [inline Include/scripts/groovy/com/kms/katalon/core/reporting/basic/reporting/ReportWriterUtil.java:207]: Can we remove this comment?

File: Include/scripts/groovy/com/kms/katalon/core/reporting/basic/reporting/ReportWriterUtil.java [hunk 2]
```diff
@@ -258,26 +314,26 @@ public static void writeJUnitReport(TestSuiteLogRecord suiteLogEntity, File logF
...
-//            public boolean shouldSkipClass(Class<?> paramClass) {
-//                return false;
-//            }
-//        };
-//        String json = new GsonBuilder().addSerializationExclusionStrategy(excludeFields)
-//                .create()
-//                .toJson(suiteLogEntity);
-//        FileUtils.writeStringToFile(new File(logFolder, "JSON_Report.json"), json, StringConstants.DF_CHARSET);
-//    }
+    // public static void writeJsonReport(TestSuiteLogRecord suiteLogEntity, File logFolder) throws IOException {
```
Comment (Minh-Chien) [inline Include/scripts/groovy/com/kms/katalon/core/reporting/basic/reporting/ReportWriterUtil.java:317]: Can we remove this comment?

File: Include/scripts/groovy/com/kms/katalon/core/reporting/basic/reporting/ReportWriterUtil.java [hunk 3]
```diff
@@ -306,13 +363,22 @@ private static void prepareHtmlContent(TestSuiteLogRecord suiteLogEntity, File d
...
-        FileUtils.writeStringToFile(destFile,
-                readFileToStringBuilder(ResourceLoader.HTML_TEMPLATE_FILE), StringConstants.DF_CHARSET);
-        FileUtils.writeStringToFile(destFile,
-                generateVars(strings, suiteLogEntity, sbModel), StringConstants.DF_CHARSET, true);
-        FileUtils.writeStringToFile(destFile,
-                readFileToStringBuilder(ResourceLoader.HTML_TEMPLATE_CONTENT), StringConstants.DF_CHARSET, true);
+        OutputStream outputStream = null;
+        Writer writer = null;
+        try {
+        	outputStream = new FileOutputStream(destFile);
```
Comment (Minh-Chien) [inline Include/scripts/groovy/com/kms/katalon/core/reporting/basic/reporting/ReportWriterUtil.java:369]: Is this correctly formatted?

File: Test Listeners/KatalonReportListener.groovy [hunk 4]
```diff
@@ -8,70 +11,68 @@ import com.kms.katalon.core.reporting.basic.reporting.ReportWriterUtil;
...
-import org.apache.commons.io.FileUtils
 
 public class KatalonReportListener {
+    public void copyDirectory(Path sourcePath, Path destinationPath) {
+        try {
+            Files.walk(sourcePath).forEach(source -> {
+                Path destination = destinationPath.resolve(sourcePath.relativize(source));
+                try {
+                    Files.copy(source, destination);
+                } catch (IOException e) {
```
Comment (Minh-Chien) [inline Test Listeners/KatalonReportListener.groovy:22]: Please help to log the exception

File: Include/scripts/groovy/com/kms/katalon/core/reporting/basic/reporting/JsStepModel.java [hunk 5]
```diff
@@ -203,8 +201,14 @@ private String getLogFolder(MessageLogRecord messageLog) {
     }
 
     private String encodeFileContent(final File file) throws FileNotFoundException, Exception, IOException {
-        try (InputStream is = new FileInputStream(file)) {
+        InputStream is = null;
```
Comment (TriLeKatalon) [inline Include/scripts/groovy/com/kms/katalon/core/reporting/basic/reporting/JsStepModel.java:204]: The old code already close the InputStream after the method has done. It's `try-with-resource`, please revert to the old code if we just want to close the stream.

File: Test Listeners/KatalonReportListener.groovy [hunk 6]
```diff
@@ -8,70 +11,68 @@ import com.kms.katalon.core.reporting.basic.reporting.ReportWriterUtil;
...
             File reportFolderFile = new File(reportFolder);
             File folderTemp = Files.createTempDirectory(reportFolderFile.getName() + "_").toFile();
-            // rename temp folder to match with report folder         
+            // rename temp folder to match with report folder
             folderTemp = Files.move(folderTemp.toPath(), folderTemp.toPath().resolveSibling(reportFolderFile.getName())).toFile();
 
-            FileUtils.copyDirectory(reportFolderFile, folderTemp);
-            String folderTempString = folderTemp.getAbsolutePath();           
-            
+            //FileUtils.copyDirectory(reportFolderFile, folderTemp);
```
Comment (TriLeKatalon) [inline Test Listeners/KatalonReportListener.groovy:50]: remove this line instead

File: Test Listeners/KatalonReportListener.groovy [hunk 7]
```diff
@@ -8,70 +11,68 @@ import com.kms.katalon.core.reporting.basic.reporting.ReportWriterUtil;
 import com.kms.katalon.core.setting.BundleSettingStore;
 import com.kms.katalon.core.util.KeywordUtil;
 import com.kms.katalon.core.util.internal.ExceptionsUtil;
-import java.nio.file.Files
-import org.apache.commons.io.FileUtils
 
 public class KatalonReportListener {
+    public void copyDirectory(Path sourcePath, Path destinationPath) {
```
Comment (TriLeKatalon) [inline Test Listeners/KatalonReportListener.groovy:16]: change this method to private if we just call it in this class only

File: Include/scripts/groovy/com/kms/katalon/core/reporting/basic/reporting/ReportWriterUtil.java [hunk 8]
```diff
@@ -306,13 +363,22 @@ private static void prepareHtmlContent(TestSuiteLogRecord suiteLogEntity, File d
...
+            writer = new OutputStreamWriter(outputStream, StringConstants.DF_CHARSET);
+            writer.write(readFileToStringBuilder(ResourceLoader.HTML_TEMPLATE_FILE));
+            generateVarsWithWriter(strings, suiteLogEntity, sbModel, writer);
+            writer.write(readFileToStringBuilder(ResourceLoader.HTML_TEMPLATE_CONTENT));
+        } finally {
+        	if (writer != null) {
```
[the remark this answers is not in the document]
Comment (linhtnlK) [reply Include/scripts/groovy/com/kms/katalon/core/reporting/basic/reporting/ReportWriterUtil.java]: Use this way to improve the image number attached on the report

File: Include/scripts/groovy/com/kms/katalon/core/reporting/basic/reporting/JsStepModel.java (hunk 5 above)
Comment (linhtnlK) [reply Include/scripts/groovy/com/kms/katalon/core/reporting/basic/reporting/JsStepModel.java:204]: Use this way to improve the image number attached on the report

**Comment by nghiadhd-2702** [comment]
@linhtnl 
I don't see you delay while generating the html report. Please delay a little bit base on chunk 
e.g we have chunk is 10, after 10 images, we shall delay 1-2 seconds before going next chunk

**Comment by linhtnlK** [comment]
> @linhtnl I don't see you delay while generating the html report. Please delay a little bit base on chunk e.g we have chunk is 10, after 10 images, we shall delay 1-2 seconds before going next chunk

I don't think it's worth it. Currently, we are able to generate more than 5000 image files (~40KB) on an HTML report, only takes around 15 sec to generate a 559MB report size, and the accepted report size is 1GB. If we apply delay while generating, it takes a long time to generate report a.

**Comment by nghiadhd-2702** [comment]
> > @linhtnl I don't see you delay while generating the html report. Please delay a little bit base on chunk e.g we have chunk is 10, after 10 images, we shall delay 1-2 seconds before going next chunk
> 
> I don't think it's worth it. Currently, we are able to generate more than 5000 image files (~40KB) on an HTML report, only takes around 15 sec to generate a 559MB report size, and the accepted report size is 1GB. If we apply delay while generating, it takes a long time to generate report a.

I know but take more time than getting out of memory and no report. 10 images is sample. A chunk with 100 images is ok.

**Review by TriLeKatalon — APPROVED** [review]

**Pull request merged**